File: ngmath/man/shgrid/shgetnp.m
.\"
.\"	$Id: shgetnp.m,v 1.4 2008-07-27 03:35:42 haley Exp $
.\"
.TH SHGETNP 3NCARG "September 1999" UNIX "NCAR GRAPHICS"
.SH NAME
SHGETNP - find the nearest points to a specified point in 3-space.
.SH SYNOPSIS
CALL SHGETNP (PX, PY, PZ, N, X, Y, Z, IFLAG, IWK, RWK, NP, IER)
.SH DESCRIPTION
.IP PX 12
(real,input) The X coordinate value for a point P 
whose nearest neighbor is to be found. 
.IP PY 12
(real,input) The Y coordinate value for a point P 
whose nearest neighbor is to be found. 
.IP PZ 12
(real,input) The Z coordinate value for a point P 
whose nearest neighbor is to be found. 
.IP X 12
(real, input) An array containing the X coordinates of the input data points.
.IP Y 12
(real, input) An array containing the Y coordinates of the input data points.
.IP Z 12
(real, input) An array containing the Z coordinates of the input data points.
.IP IFLAG 12
(integer, input) A flag that equals 0 if this is the first call 
to this subroutine for the given dataset and equals 1 otherwise. 
.IP IRK 12 
(integer, input) An integer workspace dimensioned for at least 2*N. 
.IP RWK 12
(real, input) A real workspace dimensioned for at least 11*N+6. 
.IP NP 12
(integer, output) An index such that (X(NP),Y(NP),Z(NP)) is the nearest input 
data point to P. NP = 0 if IER .NE. 0. On successive calls to this 
subroutine after the first (that is when IFLAG=1) you can find the 
Mth closest point to (PX,PY,PZ) with the Mth call. IWK and RWK should 
not be modified between calls to SHGETNP if you are wanting to
find successive nearest neighbors. 
.IP IER 12
(integer, output) An error return value.  If IER is returned as 0, then
no errors were detected. If IER is non-zero, then refer to the man
page for shgrid_errors for details.
.SH USAGE
SHGETNP is called to find the nearest point to a specified point in 3-space. 
Successive calls to SHGETNP will determine the point nearest the 
specified point exclusive of the points found in previous calls, 
i.e. successive calls can be used to find the N nearest points
for any N between one and the maximum number of points in the input dataset. 
.SH ACCESS
To use SHGETNP, load the NCAR Graphics library ngmath.
.SH SEE ALSO
shgrid,
shseti,
shgeti,
shgrid_params.
.sp
Complete documentation for Shgrid is available at URL
.br
http://ngwww.ucar.edu/ngdoc/ng/ngmath/shgrid/shhome.html
.SH COPYRIGHT
Copyright (C) 1997-2000
.br
University Corporation for Atmospheric Research
.br

The use of this Software is governed by a License Agreement.

File: ngmath/src/lib/gridpack/shgrid/shgetnp.f
C
C $Id: shgetnp.f,v 1.6 2008-07-27 03:10:13 haley Exp $
C                                                                      
C                Copyright (C)  2000
C        University Corporation for Atmospheric Research
C                All Rights Reserved
C
C The use of this Software is governed by a License Agreement.
C
      SUBROUTINE SHGETNP(PX,PY,PZ,N,X,Y,Z,IFLAG,IWK,RWK,NP,IER)
      DIMENSION X(N),Y(N),Z(N),IWK(*),RWK(*)

C
C***********************************************************
C
C  This subroutine provides a simplified interface to
C  SHGETNP3.
C
C  On input --
C
C    PX,PY,PZ = Cartesian coordinates of the point P whose
C               nearest unmarked neighbor is to be found.
C
C    N = The number of input data values.
C
C    X,Y,Z = Arrays of length N, for N .GE. 2, containing
C            the Cartesian coordinates of the input data points.
C
C    IFLAG = 0 if this is the first call to this subroutine
C            for the given dataset and = 1 otherwise. 
C
C    IWK = An integer workspace dimensioned for at least 2*N.
C
C    RWK = A real workspace dimensioned for at least 11*N+6.
C
C  On output --
C
C    NP - (X(NP),Y(NP),Z(NP)) is the nearest input data point 
C         to P.  NP = 0 if IER .NE. 0.  On successive calls to
C         this subroutine after the first (IFLAG=1) you can 
C         find the Mth closest point to (PX,PY,PZ) with the Mth 
C         call.
C
C
C  Modules required by SHGETNP3 -- none
C
C  Intrinsic functions called by SHGETNP3 -- IABS, IFIX, SQRT
C
C***********************************************************
C
C  Define defaults for the parameter values if they have not
C  been set.
C
      COMMON /SHCOMI/ NMLSTQ, NMINFL, NMCELS
C
C  Do a call forcing a BLOCKDATA to be loaded from a binary library.
C
      CALL SHBLDA
C
      IF (NMLSTQ .LT. 0) THEN
        NMLSTQ = MIN(17,N-1)
      ENDIF
C
      IF (NMINFL .LT. 0) THEN
        NMINFL = MIN(32,N-1)
      ENDIF
C
      IF (NMCELS .LT. 0) THEN
        RTMP = (REAL(N)/3.)**(0.333333)
        IF (RTMP-REAL(INT(RTMP)) .NE. 0.) THEN
          NMCELS = INT(RTMP)+1
        ELSE
          NMCELS = INT(RTMP)
        ENDIF 
      ENDIF
C
C  Initialize if this is the first call.
C
      IF (IFLAG .EQ. 0) THEN
        DO 10 I=1,N
          RWK(I) = X(I)
   10   CONTINUE
        CALL SHQSHEP(N,X,Y,Z,RWK(1),NMLSTQ,NMINFL,NMCELS,IWK(1),
     +              IWK(N+1),RWK(N+1),RWK(N+4),RMAX,
     +              RWK(N+7),RWK(2*N+7),IER)
      ENDIF
C
C  Get the index.
C
      CALL SHGETNP3(PX,PY,PZ,X,Y,Z,NMCELS,IWK(1),IWK(N+1),RWK(N+1),
     +              RWK(N+4),NP,DSQ)
C
      RETURN
      END
